onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+            IsCastTo(*consumer, TensorProto::FLOAT16)) {
+          casts.push_back(consumer);
+          continue;
+        }
+        all_float_outputs_have_casts = false;
+      }
+      require_type_change.insert(output);
+    }
+    if (has_float_outputs && all_float_outputs_have_casts && casts.size() > 0 ) {
+      VLOGS(logger, 1) << "PropagateCastsFromOutputsToInputs: Removed Cast nodes "
+                       << ConcatNames<std::vector<Node*>>(casts)
+                       << " feeding the same compute node " << node->Name();
+      RemoveCastNodes(graph, casts, removed_nodes);
+      std::unordered_set<NodeArg*> node_args;
+      for (NodeArg* input : node->MutableInputDefs()) {
+        if (IsType(*input, TensorProto::FLOAT)) {
+          node_args.insert(input);
+        }
+      }
+      InsertCastNodes(graph, node_args, true, removed_nodes);
+      ChangeTypeToFP16(require_type_change, logger);
+      VLOGS(logger, 1) << "PropagateCastsFromOutputsToInputs: Inserted Cast node to " << ConcatNames(node_args);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// Expand FP16 compute regions on the graph by example float16 compute nodes,
+// propagating float32 Cast operation down the graph and propagating float16
+// Cast operations up the graph. The following functions are performed
+// 1. Fuse subgraphs
+// 2. Propagate fp32 casts forwards
+// 3. Propagate fp16 casts back
+Status PropagateCastOps::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  // First apply the transformation to the subgraphs.
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (nullptr == node_ptr)
+      continue;  // node was removed
+
+    auto& node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+  }
+  int pass = 0;
+  bool local_modified = false;
+  do {
+    VLOGS(logger, 1) << "Propagate Cast Operations Pass " << pass << ":";
+    std::deque<onnxruntime::NodeIndex> removed_nodes;
+
+    if (local_modified) {
+      graph.Resolve();
+    }
+
+    local_modified = RemoveUnnecessaryCasts(graph, removed_nodes, logger);
+
+    // Fuse subgraphs, sibling Cast nodes with same input
+    for (auto& node: graph.Nodes()) {
+      if (std::find(removed_nodes.begin(), removed_nodes.end(), node.Index()) == removed_nodes.end()) {
+        local_modified |= FuseSiblingCasts(graph, &node, removed_nodes, logger);
+      }
+    }
+
+    // Propagate FP32 Casts forward
+    for (Node& node : graph.Nodes()) {
+      if (std::find(removed_nodes.begin(), removed_nodes.end(), node.Index()) == removed_nodes.end()) {
+        local_modified |= PropagateForwards(graph, &node, removed_nodes, level_, logger);
```

Comment:
Need multiple test cases.... 
As least one for each functions, e.g. PropagateForwards/Backward, PropagateFP16CastsFromOutputsToInputs (and reverse order), RemoveBackToBackCasts, RemoveUnnecessaryCasts, FuseSiblingCasts